src/core/components/pin/gc.js

```diff
@@ -0,0 +1,153 @@
+'use strict'
+
+const promisify = require('promisify-es6')
+const CID = require('cids')
+const base32 = require('base32.js')
+const parallel = require('async/parallel')
+const mapLimit = require('async/mapLimit')
+const { Key } = require('interface-datastore')
+const log = require('debug')('ipfs:gc')
+
+// Limit on the number of parallel block remove operations
+const BLOCK_RM_CONCURRENCY = 256
+const MFS_ROOT_DS_KEY = new Key('/local/filesroot')
+
+// Perform mark and sweep garbage collection
+module.exports = function gc (self) {
+  return promisify((callback) => {
+    const start = Date.now()
+    log(`Creating set of marked blocks`)
+
+    self._gcLock.writeLock((lockCb) => {
+      parallel([
+        // Get all blocks from the blockstore
+        (cb) => self._repo.blocks.query({ keysOnly: true }, cb),
+        // Mark all blocks that are being used
+        (cb) => createMarkedSet(self, cb)
+      ], (err, [blocks, markedSet]) => {
+        if (err) {
+          log(`Error - ${err.message}`)
+          return lockCb(err)
+        }
+
+        // Delete blocks that are not being used
+        deleteUnmarkedBlocks(self, markedSet, blocks, start, (err, res) => {
```

Comment:
Can we keep `start` local to this function and log out "complete" in the callback here instead?